src/ucp/core/ucp_request.inl

```diff
@@ -1003,19 +1003,28 @@ ucp_request_param_user_data(const ucp_request_param_t *param)
 
 static UCS_F_ALWAYS_INLINE ucs_memory_type_t
 ucp_request_get_memory_type(ucp_context_h context, const void *address,
-                            size_t length, const ucp_request_param_t *param)
+                            size_t length, ucp_datatype_t dt,
+                            const ucp_request_param_t *param)
 {
     ucp_memory_info_t mem_info;
+    const ucp_dt_iov_t *iov;
 
-    if (!(param->op_attr_mask & UCP_OP_ATTR_FIELD_MEMORY_TYPE) ||
-        (param->memory_type == UCS_MEMORY_TYPE_UNKNOWN)) {
+    if ((param->op_attr_mask & UCP_OP_ATTR_FIELD_MEMORY_TYPE) &&
+        (param->memory_type != UCS_MEMORY_TYPE_UNKNOWN)) {
+        return param->memory_type;
+    }
+
+    if ((dt & UCP_DATATYPE_CLASS_MASK) == UCP_DATATYPE_CONTIG) {
         ucp_memory_detect(context, address, length, &mem_info);
-        ucs_assert(mem_info.type < UCS_MEMORY_TYPE_UNKNOWN);
-        return (ucs_memory_type_t)mem_info.type;
+    } else if ((dt & UCP_DATATYPE_CLASS_MASK) == UCP_DATATYPE_IOV) {
+        iov = (const ucp_dt_iov_t*)address;
+        ucp_memory_detect(context, iov->buffer, iov->length, &mem_info);
```

Comment:
@brminich This assumes uniformity in iovs, right? If iov is composed of two buffers (one host, one device), then mem_info reporting mem_info.type = UCS_MEMORY_TYPE_HOST would not be accurate, right? If you agree, then mem_info will likely have to carry sys_dev, mem_type info for each vector to allow UCP to actually handle an iov with non-uniform memory types. If we don't plan to support that, shouldn't we fail here?

Reply:
maybe we can add an assert to avoid hurting performance

Reply:
If you mean, assert that all iovs have the same memory type, then agreed.

Reply:
@Akshay-Venkatesh, sorry force-pushed the changes as the original code was modified a lot